File: apps/astro/src/services/ephemeris.ts
/**
 * Ephemeris service for the astro frontend application.
 * 
 * This service provides lazy-loaded, cached access to ephemeris data
 * via the backend API, which proxies to the dedicated ephemeris server.
 */

import { useMemo } from 'react';
import { useQuery, useMutation, useQueryClient, type UseQueryResult, type UseMutationResult } from '@tanstack/react-query';
import {
  createEphemerisClient,
  type EphemerisConfig,
  type PlanetPosition,
  type PlanetName,
  dateToJulianDay,
  SUPPORTED_PLANETS
} from '@cosmichub/integrations';
import type {
  CalculationResponse,
  BatchCalculationResponse,
  EphemerisHealthResponse
} from '@cosmichub/integrations';

// Configuration for the ephemeris client
const getEphemerisConfig = (): EphemerisConfig => ({
  // Use the backend API base (frontend talks to backend, which proxies ephemeris)
  apiBaseUrl: import.meta.env.VITE_API_URL ?? 'http://localhost:8000',
  timeout: 30000,
  // API key handled by backend authentication, not needed here
});

// Query keys for React Query
export const ephemerisKeys = {
  all: ['ephemeris'] as const,
  health: () => [...ephemerisKeys.all, 'health'] as const,
  planets: () => [...ephemerisKeys.all, 'planets'] as const,
  calculation: (julianDay: number, planet: string) => 
    [...ephemerisKeys.all, 'calculation', julianDay, planet] as const,
  allPositions: (julianDay: number) => 
    [...ephemerisKeys.all, 'positions', julianDay] as const,
};

/**
 * Hook to get a configured ephemeris client instance
 */
export const useEphemerisClient = (): ReturnType<typeof createEphemerisClient> =>
  useMemo(() => createEphemerisClient(getEphemerisConfig()), []);

/**
 * Hook to check ephemeris service health
 */
export const useEphemerisHealth = (): UseQueryResult<EphemerisHealthResponse, Error> => {
  const client = useEphemerisClient();
  return useQuery<EphemerisHealthResponse, Error>({
    queryKey: ephemerisKeys.health(),
    queryFn: () => client.healthCheck(),
    staleTime: 5 * 60 * 1000, // 5 minutes
    retry: 3,
    retryDelay: (attemptIndex) => Math.min(1000 * 2 ** attemptIndex, 30000)
  });
};

/**
 * Hook to get supported planets list
 */
export const useSupportedPlanets = (): UseQueryResult<PlanetName[], Error> => {
  const client = useEphemerisClient();
  return useQuery<PlanetName[], Error>({
    queryKey: ephemerisKeys.planets(),
    queryFn: () => client.getSupportedPlanets(),
    staleTime: 24 * 60 * 60 * 1000, // 24 hours - rarely changes
    gcTime: 24 * 60 * 60 * 1000
  });
};

/**
 * Hook to calculate planetary position for a specific date and planet
 */
export const usePlanetaryPosition = (
  date: Date | null,
  planet: PlanetName,
  options?: { enabled?: boolean }
): UseQueryResult<CalculationResponse, Error> => {
  const client = useEphemerisClient();
  const julianDay = date !== null && date !== undefined ? dateToJulianDay(date) : null;
  const isEnabled = options?.enabled !== false && julianDay !== null && julianDay !== undefined;
  return useQuery<CalculationResponse, Error>({
  queryKey: julianDay != null ? ephemerisKeys.calculation(julianDay, planet) : [],
    queryFn: () => {
      if (julianDay == null) throw new Error('Date is required');
      return client.calculatePosition(julianDay, planet);
    },
    enabled: isEnabled,
    staleTime: 60 * 60 * 1000,
    gcTime: 24 * 60 * 60 * 1000
  });
};

/**
 * Hook to get all planetary positions for a specific date
 */
export const useAllPlanetaryPositions = (
  date: Date | null,
  options?: { enabled?: boolean; planets?: PlanetName[] }
): UseQueryResult<Record<PlanetName, PlanetPosition>, Error> => {
  const client = useEphemerisClient();
  const julianDay = date != null ? dateToJulianDay(date) : null;
  const hasCustomPlanets = Array.isArray(options?.planets) && options?.planets !== undefined && options.planets.length > 0;
  const planetsToCalculate: PlanetName[] = hasCustomPlanets
    ? [...new Set(options!.planets!)] // dedupe if caller passes duplicates
    : [...SUPPORTED_PLANETS]; // spread to mutable array
  const isEnabled = options?.enabled !== false && julianDay != null;
  return useQuery<Record<PlanetName, PlanetPosition>, Error>({
  queryKey: julianDay != null ? ephemerisKeys.allPositions(julianDay) : [],
    queryFn: async () => {
      if (julianDay == null) throw new Error('Date is required');
      const calculations = planetsToCalculate.map((planet) => ({
        julian_day: julianDay,
        planet
      }));
      const response = await client.calculateBatchPositions(calculations);
      const positions: Partial<Record<PlanetName, PlanetPosition>> = {};
      for (const result of response.results) {
        const planetName = result.planet as PlanetName;
        positions[planetName] = result.position;
      }
      return positions as Record<PlanetName, PlanetPosition>;
    },
    enabled: isEnabled,
    staleTime: 60 * 60 * 1000,
    gcTime: 24 * 60 * 60 * 1000
  });
};

/**
 * Hook to batch calculate multiple planetary positions
 */
export const useBatchPlanetaryCalculation = (): UseMutationResult<
  BatchCalculationResponse,
  Error,
  Array<{ date: Date; planet: PlanetName }>,
  unknown
> => {
  const client = useEphemerisClient();
  const queryClient = useQueryClient();
  return useMutation<BatchCalculationResponse, Error, Array<{ date: Date; planet: PlanetName }>>({
    mutationFn: async (requests) => {
      const calculations = requests.map((req) => ({
        julian_day: dateToJulianDay(req.date),
        planet: req.planet
      }));
      return client.calculateBatchPositions(calculations);
    },
    onSuccess: (data, variables) => {
      data.results.forEach((result, index) => {
  const request = variables[index];
  if (request !== null && request !== undefined) {
          const julianDay = dateToJulianDay(request.date);
          // Cache the full calculation response for the single planet/date
          queryClient.setQueryData(
            ephemerisKeys.calculation(julianDay, request.planet),
            result
          );
        }
      });
    }
  });
};

/**
 * Utility hook to preload planetary positions for a date range
 */
export const usePrefetchPlanetaryPositions = (): { prefetchPositions: (
  startDate: Date,
  endDate: Date,
  planets?: PlanetName[]
) => Promise<void> } => {
  const queryClient = useQueryClient();
  const client = useEphemerisClient();

  const prefetchPositions = async (
    startDate: Date,
    endDate: Date,
    planets: PlanetName[] = [...SUPPORTED_PLANETS]
  ): Promise<void> => {
    if (endDate < startDate) return; // no-op on invalid range
    const uniquePlanets = [...new Set(planets)];
    const dates: Date[] = [];
    const currentDate = new Date(startDate);
    while (currentDate <= endDate) {
      dates.push(new Date(currentDate));
      currentDate.setDate(currentDate.getDate() + 1);
    }
    const batchSize = 10;
    for (let i = 0; i < dates.length; i += batchSize) {
      const batch = dates.slice(i, i + batchSize);
      await Promise.all(
        batch.map((date) => {
          const julianDay = dateToJulianDay(date);
          return queryClient.prefetchQuery({
            queryKey: ephemerisKeys.allPositions(julianDay),
            queryFn: () => client.getAllPlanetaryPositions(julianDay),
            staleTime: 60 * 60 * 1000
          });
        })
      );
    }
  };

  return { prefetchPositions };
};

/**
 * Hook to invalidate ephemeris cache (useful for forcing refresh)
 */
export const useInvalidateEphemerisCache = (): {
  invalidateAll: () => void;
  invalidateCalculation: (julianDay: number, planet: PlanetName) => void;
  invalidateAllPositions: (julianDay: number) => void;
} => {
  const queryClient = useQueryClient();
  const invalidateAll = (): void => {
    void queryClient.invalidateQueries({ queryKey: ephemerisKeys.all });
  };
  const invalidateCalculation = (julianDay: number, planet: PlanetName): void => {
    void queryClient.invalidateQueries({
      queryKey: ephemerisKeys.calculation(julianDay, planet)
    });
  };
  const invalidateAllPositions = (julianDay: number): void => {
    void queryClient.invalidateQueries({
      queryKey: ephemerisKeys.allPositions(julianDay)
    });
  };
  return { invalidateAll, invalidateCalculation, invalidateAllPositions };
};

// Export utility functions for direct use
export {
  dateToJulianDay,
  julianDayToDate,
  formatPlanetPosition,
  getAstrologicalSign,
  SUPPORTED_PLANETS,
} from '@cosmichub/integrations';

// Export types for convenience
export type {
  PlanetPosition,
  PlanetName,
  CalculationResponse,
  BatchCalculationResponse,
  EphemerisHealthResponse,
} from '@cosmichub/integrations';

File: backend/api/routers/ephemeris.py
"""
Ephemeris API router for proxying requests to the dedicated ephemeris server.

This router provides endpoints for the frontend to access ephemeris calculations
through the backend, which then forwards requests to the ephemeris microservice.
"""

import logging
from typing import Any, Dict, List

from fastapi import APIRouter, HTTPException, status

from utils.ephemeris_client import EphemerisClient, EphemerisClientError

from ..models.ephemeris import (
    BatchCalculationRequest,
    BatchCalculationResponse,
    CalculationRequest,
    CalculationResponse,
    EphemerisHealthResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ephemeris", tags=["ephemeris"])


@router.get("/health", response_model=EphemerisHealthResponse)
async def health_check():
    """
    Check ephemeris service health.

    Returns:
        Health status of the ephemeris service
    """
    try:
        async with EphemerisClient() as client:
            health_data = await client.health_check()
            return EphemerisHealthResponse(**health_data)
    except EphemerisClientError as e:
        logger.error(f"Ephemeris health check failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Ephemeris service unavailable: {str(e)}",
        )
    except Exception as e:
        logger.error(f"Unexpected error in health check: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


@router.post("/calculate", response_model=CalculationResponse)
async def calculate_position(request: CalculationRequest):
    """
    Calculate planetary position for given Julian Day and planet.

    Args:
        request: Calculation request with julian_day and planet

    Returns:
        Calculation response with position data
    """
    try:
        async with EphemerisClient() as client:
            result = await client.calculate_position(
                request.julian_day, request.planet
            )
            return CalculationResponse(**result.model_dump())
    except EphemerisClientError as e:
        logger.error(f"Ephemeris calculation failed: {e}")
        if "Invalid planet" in str(e):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
            )
        else:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=f"Ephemeris service error: {str(e)}",
            )
    except Exception as e:
        logger.error(f"Unexpected error in calculation: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


@router.post("/calculate/batch", response_model=BatchCalculationResponse)
async def calculate_batch_positions(request: BatchCalculationRequest):
    """
    Calculate multiple planetary positions in batch.

    Args:
        request: Batch calculation request with list of calculations

    Returns:
        Batch calculation response with results
    """
    try:
        async with EphemerisClient() as client:
            # Convert Pydantic models to client models - use import directly
            from utils.ephemeris_client import (
                CalculationRequest as ClientCalculationRequest,
            )

            calculations = [
                ClientCalculationRequest(
                    julian_day=calc.julian_day, planet=calc.planet
                )
                for calc in request.calculations
            ]

            result = await client.calculate_batch_positions(calculations)
            return BatchCalculationResponse(**result.model_dump())
    except EphemerisClientError as e:
        logger.error(f"Ephemeris batch calculation failed: {e}")
        if "exceeds maximum" in str(e):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
            )
        else:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=f"Ephemeris service error: {str(e)}",
            )
    except Exception as e:
        logger.error(f"Unexpected error in batch calculation: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


@router.get("/planets", response_model=List[str])
async def get_supported_planets():
    """
    Get list of supported planets/celestial bodies.

    Returns:
        List of supported planet names
    """
    try:
        async with EphemerisClient() as client:
            planets = await client.get_supported_planets()
            return planets
    except EphemerisClientError as e:
        logger.error(f"Failed to get supported planets: {e}")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Ephemeris service error: {str(e)}",
        )
    except Exception as e:
        logger.error(f"Unexpected error getting planets: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


@router.get("/positions/{julian_day}", response_model=Dict[str, Any])
async def get_all_planetary_positions(julian_day: float) -> Dict[str, Any]:
    """
    Get all planetary positions for a specific Julian Day.

    Args:
        julian_day: Julian Day Number

    Returns:
        Dictionary with planet names as keys and position data as values
    """
    try:
        async with EphemerisClient() as client:
            # Get list of supported planets
            planets = await client.get_supported_planets()

            # Create batch request for all planets
            from utils.ephemeris_client import (
                CalculationRequest as ClientCalculationRequest,
            )

            calculations = [
                ClientCalculationRequest(julian_day=julian_day, planet=planet)
                for planet in planets
            ]

            batch_result = await client.calculate_batch_positions(calculations)

            # Convert to dictionary format
            positions: Dict[str, Dict[str, Any]] = {}
            for result in batch_result.results:
                positions[result.planet] = {
                    "position": result.position.position,
                    "retrograde": result.position.retrograde,
                }

            return_data: Dict[str, Any] = {
                "julian_day": julian_day,
                "positions": positions,
                "calculation_time": batch_result.calculation_time,
            }

            return return_data
    except EphemerisClientError as e:
        logger.error(f"Failed to get all planetary positions: {e}")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Ephemeris service error: {str(e)}",
        )
    except Exception as e:
        logger.error(f"Unexpected error getting all positions: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )
